pkg/daemon/update.go

```diff
@@ -382,29 +382,42 @@ func (dn *CoreOSDaemon) applyOSChanges(mcDiff machineConfigDiff, oldConfig, newC
 		dn.nodeWriter.Eventf(corev1.EventTypeNormal, "OSUpdateStarted", mcDiff.osChangesString())
 	}
 
+	// We previously did not emit this event when kargs changed, so we still don't
 	if mcDiff.osUpdate || mcDiff.extensions || mcDiff.kernelType {
 		// We emitted this event before, so keep it
 		if dn.nodeWriter != nil {
 			dn.nodeWriter.Eventf(corev1.EventTypeNormal, "InClusterUpgrade", fmt.Sprintf("Updating from oscontainer %s", newConfig.Spec.OSImageURL))
 		}
-
 	}
 
-	// The steps from here on are different depending on the image type, so check the image type
-	isLayeredImage, err := dn.NodeUpdaterClient.IsBootableImage(newConfig.Spec.OSImageURL)
-	if err != nil {
-		return fmt.Errorf("Error checking type of update image: %w", err)
-	}
+	// Only check the image type and excute OS changes if:
+	// - machineconfig changed
+	// - we're staying on a realtime kernel ( need to run rpm-ostree update )
+	// - we have extensions ( need to run rpm-ostree update )
+	// We have at least one customer that removes the pull secret from the cluster to "shrinkwrap" it for distribution and we want
+	// to make sure we don't break that use case, but realtime kernel update and extensions update always ran
+	// if they were in use, so we also need to preserve that behavior.
+	// https://issues.redhat.com/browse/OCPBUGS-4049
+	if mcDiff.osUpdate || mcDiff.extensions || mcDiff.kernelType || mcDiff.kargs ||
```

Comment:
No, we don't really need to. That was more "I didn't refactor that" rather than "I explicitly decided we need to check image type for kargs". Because yes, this is duplicated between both paths and it's identical: 
```
	// Apply kargs
	if mcDiff.kargs {
		if err := dn.updateKernelArguments(oldConfig.Spec.KernelArguments, newConfig.Spec.KernelArguments); err != nil {
			return err
		}
	}
```
And I probably could have factored it out of both and put it at the bottom where it wasn't dependent on the choice to make this more efficient.

Reply:
got it. It's fine to leave things the way it is, my question was mainly to see if this adds any additional benefit for the bug. We will anyway get rid-of this legacy code. sooner.